### ISSUE TITLE
Perper fails when the number of agents increases

### ISSUE THREAD

**Issue by IoanStoianov** [obecto/perper#18]
I tried simple example of creating 10 agents (in the testbet) which should return their number as a message and Perper failed.

**Comment by branimirangelov** [comment]
This issue occurs in Apocryph project (https://github.com/comrade-coop/apocryph) with the following source code:

*ManyAgentsTest.cs*
```csharp
﻿using Apocryph.Agents.Testbed;
using Apocryph.Agents.Testbed.Api;
using Microsoft.Azure.WebJobs;
using Perper.WebJobs.Extensions.Config;
using Perper.WebJobs.Extensions.Model;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Wetonomy.FunctionApp
{
	public class ManyAgentsTest
    {
        public Task<AgentContext<object>> Run(object state, AgentCapability self, object message)
        {
            var context = new AgentContext<object>(state, self);
            if (message is AgentRootInitMessage rootInitMessage)
            {
                for (int i = 0; i < 10; i++)
                {
                    context.CreateAgent("Agent" + i, "SingleAgent", new NumberMessage() { Creator = self, Number = i }, null);
                }
            }
            else if (message is NotifyNumber msg)
            {
                Console.WriteLine("{0}", msg.Number);
            }
            return Task.FromResult(context);
        }
    }

    public class NumberMessage
    {
        public int Number { get; set; }

        public AgentCapability Creator { get; set; }
    }

    public class NotifyNumber
    {
        public int Number { get; set; }
    }

    public class AgentOneWrapper
    {
        private readonly Testbed _testbed;

        public AgentOneWrapper(Testbed testbed)
        {
            _testbed = testbed;
        }

        [FunctionName("ManyAgentsTest")]
        public async Task AgentOne(
            [PerperStreamTrigger] PerperStreamContext context,
            [Perper("agentId")] string agentId,
            [Perper("initMessage")] object initMessage,
            [PerperStream("commands")] IAsyncEnumerable<AgentCommands> commands,
            [PerperStream("output")] IAsyncCollector<AgentCommands> output,
            CancellationToken cancellationToken)
        {
            await _testbed.Agent(new ManyAgentsTest().Run, agentId, initMessage, commands, output, cancellationToken);
        }
    }
}
```

*SingleAgent.cs*
```csharp
﻿using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Apocryph.Agents.Testbed;
using Apocryph.Agents.Testbed.Api;
using Microsoft.Azure.WebJobs;
using Perper.WebJobs.Extensions.Config;
using Perper.WebJobs.Extensions.Model;

namespace Wetonomy.FunctionApp
{
    public class SingleAgent
    {
        public Task<AgentContext<object>> Run(object state, AgentCapability self, object message)
        {
            var context = new AgentContext<object>(state, self);
            if (message is NumberMessage msg)
            {
                context.SendMessage(msg.Creator, new NotifyNumber
                {
                    Number = msg.Number
                }, null);
            }
                
            
            return Task.FromResult(context);
        }
    }

    public class SingleAgentWrapper
    {
        private readonly Testbed _testbed;

        public SingleAgentWrapper(Testbed testbed)
        {
            _testbed = testbed;
        }

        [FunctionName("SingleAgent")]
        public async Task AgentOne(
            [PerperStreamTrigger] PerperStreamContext context,
            [Perper("agentId")] string agentId,
            [Perper("initMessage")] object initMessage,
            [PerperStream("commands")] IAsyncEnumerable<AgentCommands> commands,
            [PerperStream("output")] IAsyncCollector<AgentCommands> output,
            CancellationToken cancellationToken)
        {
            await _testbed.Agent(new SingleAgent().Run, agentId, initMessage, commands, output, cancellationToken);
        }
    }
}
```

**Comment by branimirangelov** [comment]
In multi-threaded environment Apache Ignite thin client hangs on *PutAsync* in *PerperFabricData.AddStreamItemAsync* method and causes cluster degradation:

````
SEVERE: Critical system error detected. Will be handled accordingly to configured handler [hnd=StopNodeOrHaltFailureHandler [tryStop=false, timeout=0, super=AbstractFailureHandler [ignoredFailureTypes=[SYSTEM_WORKER_BLOCKED, SYSTEM_CRITICAL_OPERATION_TIMEOUT]]], failureCtx=FailureContext [type=SYSTEM_WORKER_BLOCKED, err=class o.a.i.IgniteException: GridWorker [name=partition-exchanger, igniteInstanceName=null, finished=false, heartbeatTs=1587144774879]]]
class org.apache.ignite.IgniteException: GridWorker [name=partition-exchanger, igniteInstanceName=null, finished=false, heartbeatTs=1587144774879]
        at org.apache.ignite.internal.IgnitionEx$IgniteNamedInstance$2.apply(IgnitionEx.java:1831)
        at org.apache.ignite.internal.IgnitionEx$IgniteNamedInstance$2.apply(IgnitionEx.java:1826)
        at org.apache.ignite.internal.worker.WorkersRegistry.onIdle(WorkersRegistry.java:233)
        at org.apache.ignite.internal.util.worker.GridWorker.onIdle(GridWorker.java:297)
        at org.apache.ignite.spi.discovery.tcp.ServerImpl$RingMessageWorker.lambda$new$0(ServerImpl.java:2663)
        at org.apache.ignite.spi.discovery.tcp.ServerImpl$MessageWorker.body(ServerImpl.java:7181)
        at org.apache.ignite.spi.discovery.tcp.ServerImpl$RingMessageWorker.body(ServerImpl.java:2700)
        at org.apache.ignite.internal.util.worker.GridWorker.run(GridWorker.java:120)
        at org.apache.ignite.spi.discovery.tcp.ServerImpl$MessageWorkerThread.body(ServerImpl.java:7119)
        at org.apache.ignite.spi.IgniteSpiThread.run(IgniteSpiThread.java:62)
````

**Comment by branimirangelov** [comment]
Additional information on the concurrent environment conditions (1 and 2 happens in parallel):

1. Single Agent A1 is outputs item
2. Single Agent A2 is executed

**Comment by branimirangelov** [comment]
Fixed in d70ae7297c1f94f66dbcd12c00bd22604094663e  (0.3.8)

**Comment by branimirangelov** [comment]
Similar behavior is still present when debugging

**Comment by branimirangelov** [comment]
Fixed in 0.3.9  (#21) - not related with multi-threading.
